control-plane-agent/api/proxy/proxy.go

```diff
@@ -0,0 +1,287 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024 Intel Corporation
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+package proxy
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+
+	pb "control-plane-agent/api/proxy/proto/mediaproxy"
+	"control-plane-agent/internal/event"
+	"control-plane-agent/internal/model"
+	"control-plane-agent/internal/registry"
+	"control-plane-agent/internal/telemetry"
+)
+
+type Config struct {
+	ListenPort int
+}
+
+type API struct {
+	pb.UnimplementedProxyAPIServer
+
+	cfg Config
+	srv *grpc.Server
+
+	requestMap   map[string]model.CommandRequest
+	requestMapMx sync.Mutex
+}
+
+func NewAPI(cfg Config) *API {
+	return &API{
+		cfg:        cfg,
+		requestMap: make(map[string]model.CommandRequest),
+	}
+}
+
+func (a *API) CancelCommandRequest(reqId string) {
+	a.requestMapMx.Lock()
+	req, ok := a.requestMap[reqId]
+	a.requestMapMx.Unlock()
+	if ok {
+		delete(a.requestMap, reqId)
+		req.Reply <- model.CommandReply{Err: errors.New("command request cancelled by initiator")}
+		close(req.Reply)
+	}
+}
+
+// RegisterMediaProxy
+func (a *API) RegisterMediaProxy(ctx context.Context, in *pb.RegisterMediaProxyRequest) (*pb.RegisterMediaProxyReply, error) {
+	if in == nil {
+		return nil, errors.New("proxy register request")
+	}
+
+	reply := event.PostEventSync(ctx, event.OnRegisterMediaProxy, map[string]interface{}{
+		"sdk_api_port":               in.SdkApiPort,
+		"st2110.dev_port_bdf":        in.St2110Config.DevPortBdf,
+		"st2110.dataplane_ip_addr":   in.St2110Config.DataplaneIpAddr,
+		"rdma.dataplane_ip_addr":     in.RdmaConfig.DataplaneIpAddr,
+		"rdma.dataplane_local_ports": in.RdmaConfig.DataplaneLocalPorts,
+	})
+	if reply.Err != nil {
+		logrus.Errorf("Proxy register req err: %v", reply.Err)
+	}
+
+	id, ok := reply.Ctx.Value(event.ParamName("proxy_id")).(string)
+	if !ok {
+		return nil, errors.New("proxy register request: registry reported no proxy id")
+	}
+
+	event.PostEventAsync(reply.Ctx, event.OnRegisterMediaProxyOk, map[string]interface{}{
+		"proxy_id": id,
+	})
+	return &pb.RegisterMediaProxyReply{ProxyId: id}, nil
+}
+
+// UnregisterMediaProxy
+func (a *API) UnregisterMediaProxy(ctx context.Context, in *pb.UnregisterMediaProxyRequest) (*pb.UnregisterMediaProxyReply, error) {
+	if in == nil {
+		return nil, errors.New("proxy unregister request")
+	}
+
+	reply := event.PostEventSync(ctx, event.OnUnregisterMediaProxy, map[string]interface{}{
+		"proxy_id": in.ProxyId,
```

Comment:
Shouldn't we check in.ProxyId for nil, not the whole in?
Or are we relying on eventual errors in reply?

Reply:
Just was wondering if we need to validate i.e. in.ProxyId before using it.